### PR TITLE
Improve hyperbahn client host port list parsing

### LINKED ISSUE #1275: Hyperbahn client hides error when failing to read hostPortFile
The only thing reported is:

```
{"log":"AssertionError: Failed to read the hostPortFile\n","stream":"stderr","time":"2015-09-04T18:04:31.825560731Z"}
```
Which isn't very helpful.

### PR DESCRIPTION
**Pull request by anson627**
To fix #1275
Host port file reading and parsing error need be exposed to users.

@Willyham @ShanniLi @Raynos

### CODE REVIEW

File: node/test/hyperbahn/advertise.js [hunk 1]
```diff
@@ -88,7 +88,7 @@ function runTests(HyperbahnCluster) {
         var client = new HyperbahnClient({
             serviceName: 'hello-bob',
             callerName: 'hello-bob-test',
-            hostPortFile: hostPortFile,
+            hostPortList: hostPortFile,
```
Comment (ShanniLi) [inline node/test/hyperbahn/advertise.js]: why this change?

**Comment by ShanniLi** [comment]
I think @Willyham meant to expose the file read error, which is different from this PR. Not sure merging hostPortFile into hostPortList is a good idea. hostPortList by its name is a list/array

**Comment by anson627** [comment]
Exposing error is part of this PR. The other part is to simplify the interface, having two parameter for the same thing make it hard to use. For example, xlate has hostPortFile in config file, if I want to overwrite it by passing --config hostPortFile [localhost:21300]. I need write code like this: if (Array.isArray(hostPortFile)), call constructor with hostPortList, otherwise, call with hostPortFile.

File: node/test/hyperbahn/advertise.js (hunk 1 above)
Comment (anson627) [inline node/test/hyperbahn/advertise.js]: this change is to simplify the constructor interface by reducing two parameters to one:
if (Array.isArray(hostPortList)) {
use it
} else if (typeof hostPortList is string) {
read and parse json file
} else {
throw error
}

As a user, I can get hostPort list from config or command line, it can be a file path or an array of strings. I just want to pass it into constructor, without worrying about which parameter to use.

**Comment by Raynos** [comment]
:-1: This is confusing.

Let's have type stability and two arguments.

**Comment by Raynos** [comment]
Please split this into two commits or two PRs.

The improved error messages are :+1:

**Comment by anson627** [comment]
Will do the split.

File: node/hyperbahn/index.js [hunk 2]
```diff
@@ -97,13 +97,13 @@ function HyperbahnClient(options) {
         self.hostPortList = options.hostPortList;
     } else if (typeof options.hostPortFile === 'string') {
         var tuple = safeSyncRead(options.hostPortFile);
-        assert(!tuple[0], 'Failed to read the hostPortFile');
+        assert(!tuple[0], 'Read host port list failed with ' + tuple[0]);
```
Comment (ShanniLi) [inline node/hyperbahn/index.js]: This is string + Error. Can you paste a sample how it will look like?

**Comment by ShanniLi** [comment]
one comment. Otherwise, lgtm.

File: node/hyperbahn/index.js (hunk 2 above)
Comment (anson627) [inline node/hyperbahn/index.js]: one sample:
Read host port list failed with Error: ENOENT, no such file or directory \'~~~~~\'
